### DISCUSSION TITLE
Reactive Panache MongoDB Always Emits On Event Loop

### DISCUSSION THREAD

**Discussion by ayhanap** [quarkusio/quarkus#30895] [Q&A]
I am using Reactive Panache MongoDB with `ReactivePanacheMongoRepository<Entity>`. My problem is; Mutiny objects like `Uni`s and `Multi`s returned from MongoDB operations like `findById()` etc. always emit on Event Loop. I usually map and transforms these reactive objects and all of my other manipulations are also happening on EventLoop. 
 This is because Unis created at https://github.com/quarkusio/quarkus/blob/main/extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/impl/Wrappers.java specifies emitsOn, which overrides default Mutiny Worker Context.
 
 Is there a way I can change this behavior without manually setting another executor on every reactive object that I get from DB operations?

[a comment by an automated account is not shown]

**Comment by loicmathieu** [comment]
This is a design choice of the MongoDB reactive client, nothing specific to MongoDB with Panache.

@cescoffier maybe you can share some light ?

**Reply by cescoffier** [reply]
In Quarkus, most if not all reactive operations must happen on the event loop. Typically it gives you propagation of data along the reactive pipeline, and so on.

**Reply by ayhanap** [reply]
Thanks for your answers, @cescoffier after writing the question I already figured this out (Which I knew but..) and I only moved one little long-running file upload operation to another executor. 

But then there is one thing `@Transactional` forces me to use another executor at the start. But other reactive objects created inside defaults to event loop again.

So right now I only have to switch to another thread to start transaction and some long running non reactive works.

**Reply by cescoffier** [reply]
Yes, that's expected. Isolate the blocking chunks (transactional or not), and run them on a worker thread pool.

**Reply by loicmathieu** [reply]
@ayhanap, MongoDB transactions are not supported by reactive MongoDB with Panache

**Reply by ayhanap** [reply]
I see, only the non reactive api supports it I guess. Do you think I can manage transaction manually with QuarkusTransaction.begin(); while still using the reactive api?

**Reply by loicmathieu** [reply]
MongoDB supports transaction, but we only integrate its support inside the synchronous client not the reactive one (this is possible, we just didn't do it yet). And there is no easy way to do it manually (a ClientSession must be pass to all calls to the client).
